src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
             /// <returns>The string matching the hash, or null.</returns>
             public static string Match(string text)
             {
-                // The longest pattern is 29 characters.
                 // The values in the array are the computed hashes of length
                 // index-1 (so runningHash[0] holds the hash for length 1).
-                var runningHash = new uint[29];
+                Span<uint> runningHash = stackalloc uint[MAX_LEN];
 
                 int longestPossiblePattern = 0;
                 for (int i = 0; i < text.Length; i++)
                 {
+                    // Notice, '-' == ('-' | 0x20)
                     uint h = text[i];
-                    if (h >= 'A' && h <= 'Z')
-                    {
-                        h |= 0x20; // ToLower
-                    }
-                    else if (!((h >= 'a' && h <= 'z') || h == '-'))
+                    h |= 0x20;
+                    if (!((uint)(h - 'a') <= 'z' - 'a' || h == '-'))
                     {
                         // If the character isn't in any of our patterns,
                         // don't bother hashing and reset the running length.
                         longestPossiblePattern = 0;
                         continue;
                     }
 
-                    for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
+                    Span<uint> rh = runningHash.Slice(0, Math.Min(i, runningHash.Length));
```

Comment:
Why doing a `Slice` here? Can't we use `runningHash` directly in the loop below?

Reply:
The change allows compiler to remove boundary checks.

Reply:
Please don't use `Slice` here, so as to make a separate fix easier. Just replacing the array with span would be a great change.

Reply:
I removed the Slice but this stopping compiler from removing boundary checks in the loop.

Reply:
We can have the Slice with comment:
```c#
// We need cut unused tail. Since 'i' is current position, the actual length is 'i + 1'. Ex., if i = 0 (first char in the string) the span length is 1, if i = 1 (second char) the span length is 2, and so on.
Span<uint> rh = runningHash.Slice(0, Math.Min(i + 1, runningHash.Length));
```

Reply:
Given that `rh[j - 1]` is used in the loop, I doubt the boundary check on that can be eliminated too.

Reply:
Yes, this works for `rh[j]` only. It is easy to exclude for `rh[j - 1]` too. Really we have no need to read previous value since we already have it in `v` variable.
I was thinking about vectorization, but I assume that you will be against the complication of the code. :-)

Since we have #18693 with tests can we return span?

Reply:
Nope, let's not using `Slice` here, just to keep the code simple and readable.
If you want, you can eliminate the call to `Math.Min` by doing `j = val1 > val2 ? val1 : val2` directly before the loop.